From eee6747f9ddf521bff6496da5c3e39c17c71c76a Mon Sep 17 00:00:00 2001
From: Jelle Licht <jlicht@fsfe.org>
Date: Sat, 20 Aug 2016 18:55:44 +0200
Subject: [PATCH] [BACKPORT] waypoint commit with both # and $ performing
 interpolation. Issue 544
To: guix-devel@gnu.org

---
 src/lexer.coffee | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/src/lexer.coffee b/src/lexer.coffee
index 6595640..d384830 100644
--- a/src/lexer.coffee
+++ b/src/lexer.coffee
@@ -118,7 +118,7 @@ exports.Lexer = class Lexer
   stringToken: ->
     return false unless starts(@chunk, '"') or starts(@chunk, "'")
     return false unless string =
-      @balancedToken(['"', '"'], ['${', '}']) or
+      @balancedToken(['"', '"'], ['${', '}'], ['#{', '}']) or
       @balancedToken ["'", "'"]
     @interpolateString string.replace STRING_NEWLINES, " \\\n"
     @line += count string, "\n"
@@ -400,7 +400,7 @@ exports.Lexer = class Lexer
           tokens.push ['IDENTIFIER', interp]
           i += group.length - 1
           pi = i + 1
-        else if (expr = @balancedString str.substring(i), [['${', '}']])
+        else if (expr = @balancedString str.substring(i), [['${', '}'], ['#{', '}']])
           tokens.push ['STRING', "$quote${ str.substring(pi, i) }$quote"] if pi < i
           inner = expr.substring(2, expr.length - 1)
           if inner.length
@@ -509,7 +509,7 @@ JS_FORBIDDEN = JS_KEYWORDS.concat RESERVED
 IDENTIFIER    = /^([a-zA-Z\$_](\w|\$)*)/
 NUMBER        = /^(((\b0(x|X)[0-9a-fA-F]+)|((\b[0-9]+(\.[0-9]+)?|\.[0-9]+)(e[+\-]?[0-9]+)?)))\b/i
 HEREDOC       = /^("{6}|'{6}|"{3}\n?([\s\S]*?)\n?([ \t]*)"{3}|'{3}\n?([\s\S]*?)\n?([ \t]*)'{3})/
-INTERPOLATION = /^\$([a-zA-Z_@]\w*(\.\w+)*)/
+INTERPOLATION = /^[$#]([a-zA-Z_@]\w*(\.\w+)*)/
 OPERATOR      = /^(-[\-=>]?|\+[+=]?|[*&|\/%=<>:!?]+)([ \t]*)/
 WHITESPACE    = /^([ \t]+)/
 COMMENT       = /^(\s*#{3}(?!#)[ \t]*\n+([\s\S]*?)[ \t]*\n+[ \t]*#{3}|(\s*#(?!##[^#])[^\n]*)+)/
-- 
2.9.3

