From 1b310a50243e37d1bace4a5fca499fc8ef218c79 Mon Sep 17 00:00:00 2001
From: Jelle Licht <jlicht@fsfe.org>
Date: Sun, 21 Aug 2016 06:55:37 +0200
Subject: [PATCH] [BACKPORT] leading then expression
To: guix-devel@gnu.org

---
 src/rewriter.coffee | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/src/rewriter.coffee b/src/rewriter.coffee
index c21e343..e6de981 100644
--- a/src/rewriter.coffee
+++ b/src/rewriter.coffee
@@ -194,6 +194,9 @@ class exports.Rewriter
   addImplicitIndentation: ->
     @scanTokens (token, i, tokens) ->
       [tag] = token
+      if tag is 'TERMINATOR' and @tag(i + 1) is 'THEN'
+        tokens.splice i, 1
+        return 0
       if tag is 'ELSE' and @tag(i - 1) isnt 'OUTDENT'
         tokens.splice i, 0, @indentation(token)...
         return 2
-- 
2.9.3

